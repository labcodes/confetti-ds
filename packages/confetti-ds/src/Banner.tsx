import React from "react";
import { isUndefined } from "lodash";

import Icon from "./Icon";
import TextButton from "./Button/TextButton";
import { IconTypes } from "./constants";
import { AbstractButtonProps } from "./Button/AbstractButton";

interface BannerProps {
  /** This is the message text string. */
  text?: string;
  /** Type of the Banner. */
  type: "info" | "warn" | "error";
  /** Sets the icon related to the bannerâ€™s message. */
  icon?: IconTypes;
  /** Object with information about the Banners's button. */
  buttonProps?: AbstractButtonProps;
}

export default function Banner({
  text,
  type = "info",
  icon,
  buttonProps = { text: "", onClick: () => {} },
}: BannerProps) {
  const renderIcon = () =>
    icon ? (
      <Icon type={icon} color="white" className="lab-banner__icon" />
    ) : undefined;

  const button = () => {
    if (buttonProps.text) {
      return type === "warn" ? (
        <TextButton size="normal" skin="dark" text={buttonProps.text} />
      ) : (
        <TextButton size="normal" skin="light" text={buttonProps.text} />
      );
    }
    return null;
  };

  const handleClick = (event) => {
    if (!isUndefined(buttonProps.onClick)) {
      buttonProps.onClick(event);
    }
  };

  return (
    <div className={`lab-banner lab-banner--${type}`}>
      {renderIcon()}
      <span className="lab-banner__message">{text}</span>
      <span
        className="lab-banner__button"
        onClick={handleClick}
        role="presentation"
      >
        {button()}
      </span>
    </div>
  );
}
