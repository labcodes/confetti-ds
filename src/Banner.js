import React from "react";
import PropTypes from "prop-types";
import { isUndefined } from "lodash";
import Icon from "./Icon";
import TextButton from "./Button/TextButton";

export default class Banner extends React.Component {
  static propTypes = {
    /** This is the message text string. */
    text: PropTypes.string.isRequired,
    /** Type of the Banner. */
    type: PropTypes.oneOf(["info", "warn", "error"]),
    /** Sets the icon related to the bannerâ€™s message. */
    icon: PropTypes.string.isRequired,
    /** Ojbect with information about the Banners's button. */
    buttonProps: PropTypes.shape({
      /** Banner's text button label. */
      text: PropTypes.string,
      /** Action to be executed when the button is clicked. */
      onClick: PropTypes.func,
    }),
  };

  static defaultProps = {
    type: "info",
    buttonProps: {
      text: "",
      onClick: () => {},
    },
  };

  icon = () => {
    const { icon } = this.props;
    return icon ? (
      <Icon type={icon} color="white" className="lab-banner__icon" />
    ) : undefined;
  };

  button = () => {
    const { buttonProps, type } = this.props;
    if (buttonProps.text) {
      return type === "warn" ? (
        <TextButton size="normal" skin="dark" text={buttonProps.text} />
      ) : (
        <TextButton size="normal" skin="light" text={buttonProps.text} />
      );
    }
    return null;
  };

  handleClick = (event) => {
    const { buttonProps } = this.props;
    if (!isUndefined(buttonProps.onClick)) {
      buttonProps.onClick(event);
    }
  };

  render() {
    const { text, type } = this.props;
    return (
      <div className={`lab-banner lab-banner--${type}`}>
        {this.icon()}
        <span className="lab-banner__message">{text}</span>
        <span
          className="lab-banner__button"
          onClick={this.handleClick}
          role="presentation"
        >
          {this.button()}
        </span>
      </div>
    );
  }
}
