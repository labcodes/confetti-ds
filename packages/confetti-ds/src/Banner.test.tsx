import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import Banner from "./Banner";

describe("Banner", () => {
  it("renders with base props", () => {
    const renderedComponent = renderer
      .create(<Banner text="Info Banner" type="info" icon="calendar" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders with button", () => {
    const renderedComponent = renderer
      .create(
        <Banner
          text="Warn Banner with button"
          type="warn"
          icon="coin"
          buttonProps={{
            text: "Warn Button",
            onClick: () => {},
          }}
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("calls props.onClick when clicked", async () => {
    const mockOnClick = jest.fn();
    const mountedBanner = mount(
      <Banner
        text="Test click on Banner"
        type="info"
        icon="arrow-left"
        buttonProps={{
          text: "Info Banner",
          onClick: mockOnClick,
        }}
      />
    );
    expect(mockOnClick.mock.calls.length).toEqual(0);
    mountedBanner.find("button").simulate("click");
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });
});
