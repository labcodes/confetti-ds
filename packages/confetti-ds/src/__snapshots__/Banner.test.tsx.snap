// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner renders with base props 1`] = `
<div
  className="lab-banner lab-banner--info"
>
  <span
    className="lab-icon lab-icon--CalendarLtr lab-icon--white lab-banner__icon"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 11a1 1 0 1 0 0-2 1 1 0 0 0 0 2Zm1 2a1 1 0 1 1-2 0 1 1 0 0 1 2 0Zm2-2a1 1 0 1 0 0-2 1 1 0 0 0 0 2Zm1 2a1 1 0 1 1-2 0 1 1 0 0 1 2 0Zm2-2a1 1 0 1 0 0-2 1 1 0 0 0 0 2Zm4-5.5A2.5 2.5 0 0 0 14.5 3h-9A2.5 2.5 0 0 0 3 5.5v9A2.5 2.5 0 0 0 5.5 17h9a2.5 2.5 0 0 0 2.5-2.5v-9ZM4 7h12v7.5c0 .83-.67 1.5-1.5 1.5h-9A1.5 1.5 0 0 1 4 14.5V7Zm1.5-3h9c.83 0 1.5.67 1.5 1.5V6H4v-.5C4 4.67 4.67 4 5.5 4Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="lab-banner__message"
  >
    Info Banner
  </span>
  <span
    className="lab-banner__button"
    onClick={[Function]}
    role="presentation"
  />
</div>
`;

exports[`Banner renders with button 1`] = `
<div
  className="lab-banner lab-banner--warn"
>
  <span
    className="lab-icon lab-icon--Money lab-icon--white lab-banner__icon"
  >
    <svg
      aria-hidden={true}
      className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="20"
      viewBox="0 0 20 20"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 9a2 2 0 1 1 4 0 2 2 0 0 1-4 0Zm2-1a1 1 0 1 0 0 2 1 1 0 0 0 0-2ZM3.5 4C2.67 4 2 4.67 2 5.5v7c0 .83.67 1.5 1.5 1.5h11c.83 0 1.5-.67 1.5-1.5v-7c0-.83-.67-1.5-1.5-1.5h-11ZM3 5.5c0-.28.22-.5.5-.5H5v1a1 1 0 0 1-1 1H3V5.5ZM3 8h1a2 2 0 0 0 2-2V5h6v1c0 1.1.9 2 2 2h1v2h-1a2 2 0 0 0-2 2v1H6v-1a2 2 0 0 0-2-2H3V8Zm10-3h1.5c.28 0 .5.22.5.5V7h-1a1 1 0 0 1-1-1V5Zm2 6v1.5a.5.5 0 0 1-.5.5H13v-1a1 1 0 0 1 1-1h1ZM5 13H3.5a.5.5 0 0 1-.5-.5V11h1a1 1 0 0 1 1 1v1Zm12-.5a2.5 2.5 0 0 1-2.5 2.5H4.09c.2.58.76 1 1.41 1h9a3.5 3.5 0 0 0 3.5-3.5v-5c0-.65-.42-1.2-1-1.41v6.41Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="lab-banner__message"
  >
    Warn Banner with button
  </span>
  <span
    className="lab-banner__button"
    onClick={[Function]}
    role="presentation"
  >
    <button
      className="lab-btn lab-btn--text lab-btn--normal lab-btn--dark"
      onClick={[Function]}
      type="button"
    >
      Warn Button
    </button>
  </span>
</div>
`;
